leif/parse_slo_rules.go

```diff
@@ -0,0 +1,201 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package leif
+
+import (
+	"encoding/json"
+	"errors"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+var oneSec, _ = time.ParseDuration("1s")
+var dayReg = regexp.MustCompile(`[0-9]+d`)
+
+type stringOrArray []string
+
+func (soa *stringOrArray) UnmarshalJSON(data []byte) error {
+	var tempInterface interface{}
+	var slice []string
+
+	err := json.Unmarshal(data, &tempInterface)
+	if err != nil {
+		//this never returns an error?
+		return err
+	}
+
+	str, isString := tempInterface.(string)
+
+	if isString {
+		slice = append(slice, str)
+	} else {
+		err = json.Unmarshal(data, &slice)
+	}
+
+	*soa = slice
+
+	return err
+}
+
+type duration time.Duration
+
+func (d duration) MarshalJSON() ([]byte, error) {
+	return json.Marshal(time.Duration(d).String())
+}
```

Comment:
Yes, in that case, it actually returns "48h0m0s". Also unsure of how important that is - these should be equivalent values to us. A workaround might be to parse this in a separate function instead of directly when unmarshalling.

Reply:
It would be possible to instead customize this marshalling function to keep formats consistent

e.g.

If duration < 1 Day => use standard json.Marshal
else if durationInHours % 24 == 0 => marshal as `(durationInHours/24)d`
else => attempt to marshal as just `hours`

?

I might be missing some edge cases. This also might not be the end of the world as we should not be Marshalling Back to JSON as a part of `leif`

Reply:
Oh yes, we could do that! However in that case unmarshalling "48h" and then immediately re-marshaling it would return "2d", right? Not sure if that would also be a concern.

Reply:
Thinking about this more.... It might just be best to annotate this with a comment explaining the behaviour as opposed to attempting to write convoluted logic (especially since we wont be re-marshaling to JSON for `leif`)